### PR TITLE
[ini] Implement input layers property @open sesame 09/08 14:17

### PR DESCRIPTION
**Pull request by zhoonit**
- [ini] Implement input layers property

```
This patch implements input layers property while decoupling input
dimensions from input layers as much as possible
to make the property dumb.

**Self evaluation:**
1. Build test: [X]Passed [ ]Failed [ ]Skipped
2. Run test: [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: Jihoon Lee <jhoon.it.lee@samsung.com>

```

### CODE REVIEW

File: nntrainer/graph/graph_node.h [hunk 1]
```diff
@@ -74,7 +74,7 @@ class GraphNode {
    *
    * @return list of name of the nodes which form input connections
    */
-  virtual const std::vector<std::string> &getInputConnections() const = 0;
+  virtual const std::vector<std::string> getInputConnections() const = 0;
```
Comment (kparichay) [inline nntrainer/graph/graph_node.h:77]: I wonder why this change is needed?

Comment (zhoonit) [reply nntrainer/graph/graph_node.h:77]: returning & is only valid if we have actual std::vector<std::string> now we have wrapper's around this we can't return the reference to it.

**Review by lhs8928 — APPROVED** [review]
LGTM

File: nntrainer/layers/layer_node.cpp [hunk 2]
```diff
@@ -469,6 +518,15 @@ void LayerNode::printPreset(std::ostream &out, PrintPreset preset) {
   print(out, flags);
 }
 
+void LayerNode::resizeInputDimensions(unsigned int size) {
+  auto cur_input_dim = init_context.getInputDimensions();
```
Comment (kparichay) [inline nntrainer/layers/layer_node.cpp:522]: consider adding an assert that numInputDimensions must always be >= number of input connections.

File: nntrainer/layers/layer_node.cpp [hunk 3]
```diff
@@ -322,18 +344,33 @@ nntrainer::Layer *LayerNode::getLayer() {
...
+}
+
+void LayerNode::addInputLayers(const std::string &in_layer) {
+  auto &input_layers =
+    std::get<std::vector<props::InputLayer>>(*layer_node_props);
+  input_layers.emplace_back(in_layer);
+}
+
+void LayerNode::setInputLayers(const std::vector<std::string> &layers) {
+  auto &input_layers =
```
Comment (kparichay) [inline nntrainer/layers/layer_node.cpp:371]: consider adding an assert that numInputDimensions must always be >= number of input connections.

File: nntrainer/layers/layer_node.cpp [hunk 4]
```diff
@@ -322,18 +344,33 @@ nntrainer::Layer *LayerNode::getLayer() {
...
+    std::get<std::vector<props::InputLayer>>(*layer_node_props);
+
+  input_layers.at(idx).set(to);
+}
+
+void LayerNode::addInputLayers(const std::string &in_layer) {
+  auto &input_layers =
+    std::get<std::vector<props::InputLayer>>(*layer_node_props);
+  input_layers.emplace_back(in_layer);
```
Comment (kparichay) [inline nntrainer/layers/layer_node.cpp:367]: consider adding an assert that numInputDimensions must always be >= number of input connections.

**Review by kparichay — APPROVED** [review]
LGTM other than the minor comments

File: nntrainer/layers/layer_node.cpp (hunk 2 above)
Comment (zhoonit) [reply nntrainer/layers/layer_node.cpp:522]: This function is being removed in the future PR. I'll check this in the finalize when we have information about dimensions

File: nntrainer/layers/layer_node.cpp (hunk 3 above)
Comment (zhoonit) [reply nntrainer/layers/layer_node.cpp:371]: https://github.com/nnstreamer/nntrainer/pull/1542#discussion_r704021846

File: nntrainer/layers/layer_node.cpp (hunk 4 above)
Comment (zhoonit) [reply nntrainer/layers/layer_node.cpp:367]: https://github.com/nnstreamer/nntrainer/pull/1542#discussion_r704021846

File: nntrainer/layers/layer_node.cpp (hunk 2 above)
Comment (kparichay) [reply nntrainer/layers/layer_node.cpp:522]: cool.

Comment (zhoonit) [reply nntrainer/layers/layer_node.cpp:522]: fyi) https://github.com/nnstreamer/nntrainer/pull/1543/files#diff-596baff7d3bfcab9105e5feba23efaedfec25215a1231fd5741dcd6c220307bbR448-R453

Comment (kparichay) [reply nntrainer/layers/layer_node.cpp:522]: awesome.

File: nntrainer/graph/network_graph.cpp [hunk 5]
```diff
@@ -719,6 +719,7 @@ int NetworkGraph::initialize() {
      */
     if (!is_input_node(lnode)) {
       auto &input_layers = lnode->getInputLayers();
+      lnode->resizeInputDimensions(input_layers.size());
```
Comment (jijoongmoon) [inline nntrainer/graph/network_graph.cpp:722]: We are not concerning when the input_layers have more than one outputs.

File: nntrainer/layers/layer_node.cpp [hunk 6]
```diff
@@ -433,6 +470,18 @@ float LayerNode::getLoss() const {
   return *loss;
 }
 
+void LayerNode::setInputDimension(const TensorDim &dim, unsigned int idx) {
+  NNTR_THROW_IF(idx >= getNumInputs(), std::out_of_range)
+    << "Setting dimensions out of bounds, idx: " << idx
```
Comment (jijoongmoon) [inline nntrainer/layers/layer_node.cpp:475]: Just for the double check, getNumInputs is equal to the number of connections for now.

**Review by jijoongmoon — APPROVED** [review]
LGTM with minor comments.

File: nntrainer/layers/layer_node.cpp (hunk 6 above)
Comment (zhoonit) [reply nntrainer/layers/layer_node.cpp:475]: Yes, this is true :)

File: nntrainer/graph/network_graph.cpp (hunk 5 above)
Comment (zhoonit) [reply nntrainer/graph/network_graph.cpp:722]: Yes. we don't have/support such case for now.
